### PR TITLE
Add alert for high memory utilization

### PR DESCRIPTION
**Pull request by andreyod**
Signed-off-by: Andrey Odarenko <andreyo@il.ibm.com>

**What this PR does / why we need it**:
This PR introduces new Prometheus alert that will fire when container that runs a VM is close to it's memory limit and close to a risky state of getting killed by runtime. Alert is firing when memory utilization is 20MB close to the memory limit.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by andreyod** [comment]
@xpivarc @ashleyschuett can you review?

**Review by jean-edouard — COMMENTED** [review]
Would it be feasible to add a unit test for that?
I see there's already a `crds_test.go`.

**Comment by andreyod** [comment]
> Would it be feasible to add a unit test for that?
> I see there's already a `crds_test.go`.

The function NewPrometheusRuleCR is covered in kubevirt_test.go and the new rule is validated and tested in hack/prom-rule-ci/prom-rules-tests.yaml.  Not sure it's feasible to add a unit test for specific rule.

**Review by jean-edouard — APPROVED** [review]

**Comment by vladikr** [comment]
@andreyod Jira issues are not accessible to anyone outside of RH/IBM - we should only mention publicly accessible links.

**Comment by sradco** [comment]
@andreyod The alert rule name it doesn't comply with the alert names and should be with a kubevirt_ prefix.
kubevirt_vmi_high_memory_usage or kubevirt_vm_high_memory_usage
There is a PR in progress for adding the kubevirt_ prefix to the other alert rules: https://github.com/kubevirt/kubevirt/pull/4636

File: pkg/virt-operator/resource/generate/components/crds.go [hunk 1]
```diff
@@ -766,6 +766,18 @@ func NewPrometheusRuleSpec(ns string, workloadUpdatesEnabled bool) *promv1.Prome
 							"summary": "More than 80% of the rest calls failed in virt-handler for the last 5 minutes",
 						},
 					},
+					{
+						Alert: "KubevirtVmHighMemoryUsage",
+						Expr:  intstr.FromString("sum by(pod, container) ( kube_pod_container_resource_limits_memory_bytes{pod=~'virt-launcher-.*', container='compute'} - on(pod,container) container_memory_working_set_bytes{pod=~'virt-launcher-.*', container='compute'}) < 20971520"),
```
Comment (sradco) [inline pkg/virt-operator/resource/generate/components/crds.go:771]: @andreyod I believe we should create a recording rule for this calculation.
The advantage of the recording rule is that it allow you to precompute frequently needed or computationally expensive expressions and save their result as a new set of time series.

Comment (andreyod) [reply pkg/virt-operator/resource/generate/components/crds.go]: Done

File: pkg/virt-operator/resource/generate/components/crds.go [hunk 2]
```diff
@@ -766,6 +766,22 @@ func NewPrometheusRuleSpec(ns string, workloadUpdatesEnabled bool) *promv1.Prome
...
+					{
+						Record: "kubevirt_vm_container_free_memory_bytes",
+						Expr:   intstr.FromString("sum by(pod, container) ( kube_pod_container_resource_limits_memory_bytes{pod=~'virt-launcher-.*', container='compute'} - on(pod,container) container_memory_working_set_bytes{pod=~'virt-launcher-.*', container='compute'})"),
+					},
+					{
+						Alert: "KubevirtVmHighMemoryUsage",
+						Expr:  intstr.FromString("kubevirt_vm_container_free_memory_bytes < 20971520"),
+						For:   "1m",
+						Annotations: map[string]string{
+							"description": "Container {{ $labels.container }} in pod {{ $labels.pod }} memory utilization is 20 MB close to the limit",
```
Comment (sradco) [inline pkg/virt-operator/resource/generate/components/crds.go:778]: @andreyod I think in would be more understandable if we write:
Container {{ $labels.container }} in pod {{ $labels.pod }} free memory is under 20 MB and is close to its limit

Comment (andreyod) [reply pkg/virt-operator/resource/generate/components/crds.go:778]: Maybe:
Container {{ $labels.container }} in pod {{ $labels.pod }} free memory is less than 20 MB and it is close to memory limit

Comment (sradco) [reply pkg/virt-operator/resource/generate/components/crds.go:778]: +1

Comment (andreyod) [reply pkg/virt-operator/resource/generate/components/crds.go]: Changed

**Pull request merged**